### ISSUE TITLE
Some more possible C API issues / suggestions

### ISSUE THREAD

**Issue by aseyboldt** [roualdes/bridgestan#103]
While working on rust langue support I found a few more (somewhat nitpicky) issues with the C API:

## File paths

`bs_construct` takes a file path as one input (as a `const *char`). It then passes that to `std::ifstream`. I spend quite some time yesterday trying to figure out how different operating systems handle encodings for filenames and paths, and it seems to be quite a mess. I'm really not sure how I should pass in the path on different operating systems so that it works with non ASCII paths.
Is that really worth the effort to figure that out? I think it would be easier to specify that the input should always be a utf8 encoded json c-string. (Or change it to `const *char` + strlen?).
Language bindings can then use their languages mechanism to handle file paths, so the C API doesn't have to worry about that at all.

It is also a bit unclear right now what I should do if there is no input data. I think I have to pass in an empty string, but I think a null pointer would be cleaner?

## Access to transformed data

I think there is currently no way to access quantities from the `transformed data` block? It would be nice if there was a way to extract those, so that they could for instance end up in the `data` group of an InferenceData object.

## Integer generated values

If I understand the type system of stan correctly, it has primitive types `float64` and `int32`.
In all functions of the C API that handle parameters, everything is a float64. I think that should be lossless, because every i32 value can be exactly represented as one float64 value.
I am not clear how complex values are encoded in the variable names though, and it would be great to have a way to figure out the actual types of the variables, so that I can convert them later in the traces.

**Comment by WardBrian** [comment]
> File paths

We already optionally support c-string encoded JSON. The argument name is a bit historical/misleading, but the API doc specifies:
```
 * @param[in] data_file C-style string. This is either a
 * path to JSON-encoded data file (must end with ".json"), or
 * a JSON string literal.
```

It's not stated explicitly here, but Stan uses the [`rapidjson`](https://rapidjson.org/md_doc_encoding.html) parser which should be UTF-8 aware (we pass the `kParseValidateEncodingFlag` flag)

> Access to transformed data

These are stored as members of the generated model class object and are not accessible through the `model_base` interface (or, more generally, not accessible by any code which does not know the names ahead of time).

It's a reasonable thing to request but can't be done without changes to things upstream in stan-dev/stan

> Integer generated values

This is in general a problem we have in Stan, `write_array` (which outputs parameters, transformed parameters, and generated quantities) stores its output only in a `vector<double>`. We currently don't expose any way in the Stan model to inspect the types of these things, but it will probably come one day as part of a larger refactor of how we structure outputs (see https://github.com/stan-dev/design-docs/pull/48)

So, sadly, this is also a reasonable request which is not entirely under bridgestan's control at the moment.

For complex numbers, we output them as two consecutive doubles. You can tell a variable is complex if the final index has `real` or `imag` rather than a number, for example `array[2] complex v` will have `param_names` `"v.1.real,v.1.imag,v.2.real,v2.imag"`

**Comment by WardBrian** [comment]
Oh, and

> It is also a bit unclear right now what I should do if there is no input data. I think I have to pass in an empty string, but I think a null pointer would be cleaner?

Yes, currently it is an empty string. nullptr does seem like a good option. Supporting both just requires another conditional in the constructor

**Comment by WardBrian** [comment]
>  I still think it might make sense though to just remove the file path functionality, given that (I think?) it's probably going to be implemented in an incomplete/buggy way in pretty much all language interfaces?

I'm not sure what you mean here (or at least, I don't see the reason for removing it).

> If that happens in stan as well it would be nice to have it in the C API directly, it plays pretty nice with pure C interfaces.

We definitely intend to keep BridgeStan up to date with any new functionality in the model class, so whatever CmdStan eventually uses for more structured output will most likely find its way into BridgeStan in some way or another.

**Comment by aseyboldt** [comment]
> I'm not sure what you mean here (or at least, I don't see the reason for removing it).

Maybe I'm being overly exact about this.
But the problem as I see it is that different C++ library implementations behave differently when they open an `fstream` with a `*char` (See the filename description [here](https://cplusplus.com/reference/fstream/ifstream/open/)). On unix filenames actually are just bytes, and don't have a specified encoding, although in more recent distributions paths are pretty much always UTF8 encoded in practice. At least there is a well defined thing I need to pass in to get a specific file.
But I think on Windows filenames are unicode sequences, and internally stored as utf16. So given a path (ie a sequence of characters encoded as utf16), which chars do I need to give to `bs_construct`? If the C++ compiler was MSCV (which I can't even know from the C API), I think the answer is: Encode the char sequence using the the ANSI codepage (some strange windows specific encoding), and if that is not possible because it contains eg asian characters I'm out of luck and just can't (I think). I couldn't even find docs about what would happen if the compiler is mingw...
Who would have though that opening a file is that tricky :-)

**Comment by WardBrian** [comment]
Ah, I am following now. I think that, in practice, this has not been an issue for Stan. `ifstream` is used heavily in e.g. CmdStan but to my knowledge this has not been a source of bugs/bug reports (while basically everything else Windows-y has). 

Maybe this is just lucky because we're using mingw-gcc, which I guess does something "reasonable" on Windows. As a small (N=1) set of evidence, I recently have compiled both CmdStan programs and BridgeStan using the MSVC-compliant `clang` executable and still had no issues opening files with `ifstream`. But, I have my system language set to English and I have almost exclusively ASCII folder names, so again, maybe I'm just lucky. 

In practice, if you give us a path to a file which does not exist, (either because _it actually does not exist_ or because the bytes you gave us don't agree with how the OS represented the path to the file), BridgeStan will just throw a `"Cannot read input file: ..."` error. Not ideal, obviously, but also not like we're going to run into UB or anything too crazy.

**Comment by aseyboldt** [comment]
I don't see anything worse than an "file not found" either. I've seen quite a few cases where people (mostly programming beginners, you stop doing that after you got burned a couple of times) had a "ä/ü/ß" or so in their folder or file names and ran into issues because of that. But I don't think those people are likely to then go and open github issues...

Still, clearly a minor issues though. :-)

**Comment by WardBrian** [comment]
It's definitely something worth noting. It seems like `boost` might have some more portable drop-in replacements for `ifstream` which might be worth looking into at some point. I'd like to at least keep the option of passing a path around, since it can save (sometimes pretty big) string being passed

We have many more issues with the fact that we rely so heavily on `make` for our builds and having a space in your path can send it haywire, which can be really bad if someone has their Windows username set to `FirstName LastName`, so _all_ their paths have one unless they're working outside their user folder.

**Comment by bob-carpenter** [comment]
> this has not been an issue for Stan.

I'd still rather do this "right", whatever that means.  

I think we may just have not run into this because our users aren't as unicode-happy as the Julia community and it's usually easy to rename files.

**Comment by aseyboldt** [comment]
It does look like boost might be able to at least help a bit. Another option might be to pass a file descriptor instead of a path, but I don't think that's entirely possible with just the C++ std lib either...

I don't really see a practical downside to reading data files in the language wrappers though. This would mean that there is one extra pass over the dataset (once to read it into a string, and then once to parse it in stan), but it would still be fewer passes than currently in cmdstan (once per chain).
When we actually sample something we usually have one pass over the dataset *per grad eval*, so I don't really see the point of worrying about one extra pass up front, or am I missing something?

**Comment by WardBrian** [comment]
The downside is if we punt this up to the interfaces we need to worry about it in `$NUM_INTERFACES` places, rather than the (hypothetical) advantage of "getting it right" once in the C++ code. 

Now, _most likely_, the higher-level languages get this right and handle all the caveats from above, and I'm also suspecting that an interface which is "correct", "portable", and "implementable in a pure C API" is probably a monumental task if you really want to dig into those criteria. 
So I wouldn't be too opposed to adjusting our interfaces to pass the already-read-in JSON object as a rule.

That said, I would like to keep the option in the C API, even if our interfaces don't use it. Someone might be using it _in pure C_, where the best-effort solution we provide is probably as good as whatever they'd use to read the file before passing it. Since our implementation is safe (in the sense that it will either read the file, or it will propagate an error) I don't think it's worth removing it just because it may propagate an error in some cases where a hypothetical perfect implementation would not.

**Comment by WardBrian** [comment]
I think after #112 the actionable items here are resolved.
